OlinkAnalyze/R/olink_normalization_bridgeable.R

```diff
@@ -0,0 +1,318 @@
+
+  # cleanup input datasets ----
+
+  lst_df_clean <- lapply(
+    lst_df,
+    function(l_df) {
+      l_df |>
+        dplyr::filter(
+          # only customer samples
+          .data[["SampleType"]] == "SAMPLE"
+          # remove interal control assays
+          & .data[["AssayType"]] == "assay"
+          # remove datapoints with very few counts
+          & .data[["Count"]] > .env[["min_datapoint_cnt"]]
+        ) |>
+        # keep only relevant columns (e.g. SampleID, OlinkID, NPX, Count)
+        dplyr::select(
+          dplyr::all_of(
+            c(
+              ref_cols$sample_id,
+              ref_cols$olink_id,
+              ref_cols$quant,
+              "Count"
+            )
+          )
+        )
+    }
+  )
+
+  # append the two datasets to each other ----
+
+  df_combo <- lst_df_clean |>
+    # append the datasets to each other
+    dplyr::bind_rows(
+      .id = "df_name"
+    ) |>
+    # pivot to wider so that there is one entry for each concatenated assay and
+    # sample identifier
+    tidyr::pivot_wider(
+      names_from = dplyr::all_of(
+        c("df_name")
+      ),
+      values_from = dplyr::all_of(
+        c(
+          ref_cols$quant,
+          "Count"
+        )
+      )
+    ) |>
+    # drop NA values not matching between datasets
+    # this is not intended to remove data from non-matching sample identifiers,
+    # but to remove datapoints non-matching because they were filtered out.
+    tidyr::drop_na()
+  rm(lst_df_clean)
+
+  # check range and low counts ----
+
+  # column names from each product quantification to allow computing the range
+  # difference and low median counts
+  range_names <- paste("range", ref_cols$quant, names(lst_df), sep = "_")
+  med_cnt_names <- paste("low_cnt_Count", names(lst_df), sep = "_")
+
+  df_combo <- df_combo |>
+    dplyr::group_by(
+      .data[[ref_cols$olink_id]]
+    ) |>
+    dplyr::mutate(
+      # quantification range
+      dplyr::across(
+        dplyr::starts_with(ref_cols$quant),
+        list(
+          range = ~ quantile(x = .x, probs = 0.9, names = FALSE, na.rm = TRUE) -
+            quantile(x = .x, probs = 0.1, names = FALSE, na.rm = TRUE)
+        ),
+        .names = "{fn}_{col}"
+      ),
+      # median counts
+      dplyr::across(
+        dplyr::starts_with("Count"),
+        list(
+          low_cnt = ~ median(x = .x, na.rm = TRUE) < .env[["med_cnt"]]
+        ),
+        .names = "{fn}_{col}"
+      )
+    ) |>
+    dplyr::ungroup() |>
+    # compute the actual range and the low count flag
+    dplyr::mutate(
+      range_diff = abs(.data[[range_names[1L]]] - .data[[range_names[2L]]]),
+      low_cnt = .data[[med_cnt_names[1L]]] | .data[[med_cnt_names[2L]]]
+    ) |>
+    # remove additional columns
+    dplyr::select(
+      -dplyr::all_of(
+        c(range_names, med_cnt_names)
+      )
+    )
+  rm(range_names, med_cnt_names)
+
+  # check and remove outlier samples ----
```

Comment:
I do not think that we should change the order. Range is 90th-10th percentile which excludes extreme outliers already, so it is immune to outlier samples. Also median counts is the 50th percentile of the counts, which is immune to outliers too. Outlier samples are used to prevent R2 and KS statistics to have lower values because of extreme outliers. Conceptually, the same order of operations is followed in olink_qc_plot.

Removing outliers before computing range can lead to removing relevant data points. For example, if one removes outliers outside the range [-3*IQR,3*IQR] first, and then computes range 10th-90 percentile, some samples that were originally not outliers (lied inside the the [-3*IQR,3*IQR]) are not use to compute range because they just happen to be above the 90th or below the 10th percentile. This can lead to artificially narrower range, and if repeated a few times should leave you with a handful of data points.